### PR TITLE
bcm2835-i2s: Changes for allowing asymmetric sample formats.

### PR DESCRIPTION
**Pull request by gtrainavicius**
Hey,

I've noticed that when overdubbing audio with Audacity using 16-bit sample format for recording, it still uses 32 bits for playback, causing corrupt audio to get recorded. The bcm2835-i2s forces format symmetry between the streams, but it needn't be the case. (I alternatively tried to set .symmetric_formats=1 in bcm2835_i2s_dai, but Audacity wasn't able to detect this, and failed to record completely which would be unacceptable.)

I've made some changes that allow having asymmetric sample formats, and tested various combinations of simultaneous playback and recording with the pisound card, however, I don't own any other audio cards, so I can't test the changes using them.

Anyway, I tried to keep the changes to minimum, the only cases these changes should affect is when the user software is requesting asymmetric sample format for playback and capture, in all other cases, it should behave the same as without the changes.

Thank you,
Giedrius.

### CODE REVIEW

**Comment by pelwell** [comment]
This change looks reasonable to me - it compiles, and I2S playback still works. @HiassofT @msperl Any comments?

**Comment by msperl** [comment]
more testing and upstreaming maybe?

**Comment by gtrainavicius** [comment]
Let me know if there's anything I can do to help :)

**Comment by HiassofT** [comment]
I'll run some playback/recording tests with the cirrus logic audio card
and then report back - could take a few days (busy with work)

File: sound/soc/bcm/bcm2835-i2s.c [hunk 1]
```diff
@@ -465,26 +467,46 @@ static int bcm2835_i2s_hw_params(struct snd_pcm_substream *substream,
...
+	default:
+		return -EINVAL;
+	}
 
 	/* Setup the I2S mode */
+	/* Keep existing FTXP and FRXP values. */
+	regmap_read(dev->i2s_regmap, BCM2835_I2S_MODE_A_REG, &mode);
+
+	previous_ftxp = mode & BCM2835_I2S_FTXP;
+	previous_frxp = mode & BCM2835_I2S_FRXP;
```
Comment (flatmax) [inline sound/soc/bcm/bcm2835-i2s.c:487]: If these previous variables aren't used, they should be removed.

File: sound/soc/bcm/bcm2835-i2s.c [hunk 2]
```diff
@@ -465,26 +467,46 @@ static int bcm2835_i2s_hw_params(struct snd_pcm_substream *substream,
...
+	switch (substream->stream) {
+	case SNDRV_PCM_STREAM_PLAYBACK:
+		regmap_write(dev->i2s_regmap, BCM2835_I2S_TXC_A_REG, format);
+		break;
+	case SNDRV_PCM_STREAM_CAPTURE:
+		regmap_write(dev->i2s_regmap, BCM2835_I2S_RXC_A_REG, format);
+		break;
+	default:
+		return -EINVAL;
+	}
```
Comment (flatmax) [inline sound/soc/bcm/bcm2835-i2s.c:480]: The format is the same, why split ?

File: sound/soc/bcm/bcm2835-i2s.c [hunk 3]
```diff
@@ -465,26 +467,46 @@ static int bcm2835_i2s_hw_params(struct snd_pcm_substream *substream,
...
 
 	/* Setup the I2S mode */
+	/* Keep existing FTXP and FRXP values. */
+	regmap_read(dev->i2s_regmap, BCM2835_I2S_MODE_A_REG, &mode);
```
Comment (flatmax) [inline sound/soc/bcm/bcm2835-i2s.c:484]: The implications of restoring the mode are quite large and risky! Why not let the file set it up from scratch ?
All the lines which treat mode in the file do "mode |=" to alter mode. They all assume that mode starts as : mode = 0;

**Review by flatmax — COMMENTED** [review]
There are unused variables in this patch. There is a risky introduction of restoring mode, rather then starting with it =0.

File: sound/soc/bcm/bcm2835-i2s.c [hunk 4]
```diff
@@ -465,26 +467,46 @@ static int bcm2835_i2s_hw_params(struct snd_pcm_substream *substream,
...
+	 * Retain the frame packed mode (2 channels per 32 bit word)
+	 * of the other direction stream intact. The formats of each
+	 * direction can be different as long as the frame length is
+	 * shared for both.
+	 */
+	packed = data_length <= 16;
+
+	switch (substream->stream) {
+	case SNDRV_PCM_STREAM_PLAYBACK:
+		mode |= previous_frxp;
```
Comment (gtrainavicius) [inline sound/soc/bcm/bcm2835-i2s.c:501]: This is the location where previous_... variables are used.

File: sound/soc/bcm/bcm2835-i2s.c (hunk 1 above)
Comment (gtrainavicius) [reply sound/soc/bcm/bcm2835-i2s.c:487]: They are used, see below.

File: sound/soc/bcm/bcm2835-i2s.c (hunk 3 above)
Comment (gtrainavicius) [reply sound/soc/bcm/bcm2835-i2s.c:484]: Only the stream direction specific state is kept, not the whole mode.

There's mode = 0 line still there, see below.

File: sound/soc/bcm/bcm2835-i2s.c [hunk 5]
```diff
@@ -465,26 +467,46 @@ static int bcm2835_i2s_hw_params(struct snd_pcm_substream *substream,
...
+	}
 
 	/* Setup the I2S mode */
+	/* Keep existing FTXP and FRXP values. */
+	regmap_read(dev->i2s_regmap, BCM2835_I2S_MODE_A_REG, &mode);
+
+	previous_ftxp = mode & BCM2835_I2S_FTXP;
+	previous_frxp = mode & BCM2835_I2S_FRXP;
+
 	mode = 0;
```
Comment (gtrainavicius) [inline sound/soc/bcm/bcm2835-i2s.c:489]: Here the mode is reset.

**Comment by gtrainavicius** [comment]
@flatmax It looks like you've missed some lines, the comments you have made are not correct. Please see the places I've pointed out in the code.

File: sound/soc/bcm/bcm2835-i2s.c (hunk 2 above)
Comment (gtrainavicius) [reply sound/soc/bcm/bcm2835-i2s.c:480]: hw_params is called once for each direction (playback, capture), depending on what the user software wants to do. The substream->stream stores the requested direction. So the whole idea of this pull request is to react and change only the state that is relevant to the requested direction.

**Comment by pelwell** [comment]
The limited context around each difference can make it hard to follow. I wondered whether the two separate previous_f?xp variables could be replaced with previous_mode, with the masking applied when mode is being assembled - something like:
```
	switch (substream->stream) {
	case SNDRV_PCM_STREAM_PLAYBACK:
		mode = (previous_mode & BCM2835_I2S_FRXP) |
			(packed ? BCM2835_I2S_FTXP : 0);
		break;
	case SNDRV_PCM_STREAM_CAPTURE:
		mode = (previous_mode & BCM2835_I2S_FTXP) |
			(packed ? BCM2835_I2S_FRXP : 0);
		break;
```
but I do think the patch is correct.

File: sound/soc/bcm/bcm2835-i2s.c (hunk 2 above)
Comment (flatmax) [reply sound/soc/bcm/bcm2835-i2s.c:480]: OK seems good.

File: sound/soc/bcm/bcm2835-i2s.c (hunk 3 above)
Comment (flatmax) [reply sound/soc/bcm/bcm2835-i2s.c:484]: Yep I see it now.

File: sound/soc/bcm/bcm2835-i2s.c (hunk 1 above)
Comment (flatmax) [reply sound/soc/bcm/bcm2835-i2s.c:487]: These can still be removed, see my new comment for line number 502

File: sound/soc/bcm/bcm2835-i2s.c [hunk 6]
```diff
@@ -310,6 +310,7 @@ static int bcm2835_i2s_hw_params(struct snd_pcm_substream *substream,
 	unsigned int sampling_rate = params_rate(params);
 	unsigned int data_length, data_delay, bclk_ratio;
 	unsigned int ch1pos, ch2pos, mode, format;
+	unsigned int previous_ftxp, previous_frxp;
```
Comment (flatmax) [inline sound/soc/bcm/bcm2835-i2s.c:313]: These are still unnecessary.

**Comment by gtrainavicius** [comment]
@pelwell My intention was to try to be explicit about what is kept from previous state, but I'd agree with your version too.

File: sound/soc/bcm/bcm2835-i2s.c [hunk 7]
```diff
@@ -465,26 +467,46 @@ static int bcm2835_i2s_hw_params(struct snd_pcm_substream *substream,
...
+	 * of the other direction stream intact. The formats of each
+	 * direction can be different as long as the frame length is
+	 * shared for both.
+	 */
+	packed = data_length <= 16;
+
+	switch (substream->stream) {
+	case SNDRV_PCM_STREAM_PLAYBACK:
+		mode |= previous_frxp;
+		mode |= packed ? BCM2835_I2S_FTXP : 0;
```
Comment (flatmax) [inline sound/soc/bcm/bcm2835-i2s.c:502]: Replace this with : mode &= packed ? BCM2835_I2S_FTXP : 0;

File: sound/soc/bcm/bcm2835-i2s.c [hunk 8]
```diff
@@ -465,26 +467,46 @@ static int bcm2835_i2s_hw_params(struct snd_pcm_substream *substream,
...
+		break;
+	case SNDRV_PCM_STREAM_CAPTURE:
+		mode |= previous_ftxp;
```
Comment (flatmax) [inline sound/soc/bcm/bcm2835-i2s.c:505]: Remove this line.

File: sound/soc/bcm/bcm2835-i2s.c [hunk 9]
```diff
@@ -465,26 +467,46 @@ static int bcm2835_i2s_hw_params(struct snd_pcm_substream *substream,
...
+		mode |= packed ? BCM2835_I2S_FRXP : 0;
```
Comment (flatmax) [inline sound/soc/bcm/bcm2835-i2s.c:506]: Replace this with : mode &= packed ? BCM2835_I2S_FRXP : 0;

File: sound/soc/bcm/bcm2835-i2s.c (hunk 4 above)
Comment (flatmax) [reply sound/soc/bcm/bcm2835-i2s.c:501]: Oh yep, however you can remove this line, not necessary.

File: sound/soc/bcm/bcm2835-i2s.c (hunk 5 above)
Comment (flatmax) [reply sound/soc/bcm/bcm2835-i2s.c:489]: Yep, I see, but you can get rid of this for simplification because of changes suggested from line 502

File: sound/soc/bcm/bcm2835-i2s.c [hunk 10]
```diff
@@ -465,26 +467,46 @@ static int bcm2835_i2s_hw_params(struct snd_pcm_substream *substream,
...
-		 * so the format will be the same.
-		 */
-		mode |= BCM2835_I2S_FTXP | BCM2835_I2S_FRXP;
+	/*
+	 * Retain the frame packed mode (2 channels per 32 bit word)
+	 * of the other direction stream intact. The formats of each
+	 * direction can be different as long as the frame length is
+	 * shared for both.
+	 */
+	packed = data_length <= 16;
```
Comment (flatmax) [inline sound/soc/bcm/bcm2835-i2s.c:497]: Not sure if it is good practice, but the packed variable can also be removed and this can be inline below.

**Review by flatmax — COMMENTED** [review]
I see what you are saying, however it is better to simplify this commit further - check my suggestions.

**Comment by gtrainavicius** [comment]
I'm sure it would work as you describe it, in my opinion being more explicit about the things that's happening (leaving the mode = 0; having explicit variables for what state is kept (previous_mode would work well too)) is better for readability.

Clearing most of mode variable in some switch by doing &= and keeping some of the state is a bit less explicit and easier to miss than an unconditional mode = 0;

Having more variables on stack should be really cheap on performance, it'd be great though if there was no warning for declaring the variable closer to where it's actually used rather than forcing it to be declared at the top of the function... :)

Anyway, up to you guys to decide what style to use.

**Comment by HiassofT** [comment]
I've ran some tests with the Cirrus Logic audio card (play, rec, play+rec
at 16 and 32 bit) and everything worked fine.

The wm5102 codec has symmetric_samplebits=1 so I couldn't test the
asymmetric case - but from monitoring mode via printk it seems
to work as expected.

About coding style: IMO keep it as it is. Doing acrobatics with
bit shifts and masks make the code hard to read, I'm not a huge
fan of that.

So, from my side, I think this PR is good to merge.

**Comment by pelwell** [comment]
Thanks, everyone.

**Comment by gtrainavicius** [comment]
Thank you!

**Comment by msperl** [comment]
please upstream the patch!

**Comment by flatmax** [comment]
I notice that this has been committed. Don't you think this is far simpler ?
if  (data_length <= 16)
   mode &= BCM2835_I2S_FRXP 
else 
  mode=0;

Same for the other substream.

**Comment by HiassofT** [comment]
Sorry, looks like I missed some nasty bugs when testing this - details are in issue #1799

**Comment by pelwell** [comment]
Reverted.
